File: web/src/lib/minidoc.ts
export type MiniDocSection = {
  text: string;
  concepts: string[];
};

export type WorkedExampleSection = MiniDocSection & {
  code_example: string;
  code_breakdown: string;
};

export type QuickCheck = {
  question: string;
  hint: string;
  anchor_text: string;
  anchor_section_key: "core_idea" | "mental_model" | "worked_example";
  type: "recall" | "apply" | "connect";
};

export type MiniDoc = {
  id: string;
  topic: string;
  hook: string;
  core_idea: MiniDocSection;
  mental_model: MiniDocSection;
  worked_example: WorkedExampleSection;
  quick_checks: QuickCheck[];
  mastery: number;     // 0..1
  created_at: string;  // ISO date-time
};

export type CreateDocRequest = {
  input: string;
  source: "chat" | "pdf";
  context_id: string | null;
};

export type ClarifyRequest = {
  doc_id: string;
  question: string;
};

export type ClarifyResponse = {
  doc_id: string;
  answer: string;
  summary: string;
  last_updated: string; // ISO datetime
};

File: api/app/models/minidoc.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class MiniDocSection(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    text: str
    concepts: list[str]


class WorkedExampleSection(MiniDocSection):
    code_example: str  # empty string if topic is not code-related
    code_breakdown: str  # empty string if no code


class QuickCheck(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    question: str
    hint: str
    anchor_text: str  # verbatim sentence from any section that best answers this question
    anchor_section_key: Literal["core_idea", "mental_model", "worked_example"]
    type: Literal["recall", "apply", "connect"]


class MiniDocDraft(BaseModel):
    """Shape the LLM must return. Backend injects id / mastery / created_at."""

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    topic: str
    hook: str  # 1–2 sentences: why this matters now
    core_idea: MiniDocSection
    mental_model: MiniDocSection
    worked_example: WorkedExampleSection
    quick_checks: list[QuickCheck] = Field(min_length=1)


class MiniDoc(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    id: str
    topic: str
    hook: str
    core_idea: MiniDocSection
    mental_model: MiniDocSection
    worked_example: WorkedExampleSection
    quick_checks: list[QuickCheck]
    mastery: float = Field(ge=0.0, le=1.0)
    created_at: datetime
